interfaces/builtin/content.go

```diff
@@ -169,23 +168,23 @@ func (iface *contentInterface) path(attrs interfaces.Attrer, name string) []stri
 // $SNAP_COMMON. If there are no variables then $SNAP is implicitly assumed
 // (this is the behavior that was used before the variables were supporter).
 func resolveSpecialVariable(path string, snapInfo *snap.Info) string {
-	if strings.HasPrefix(path, "$SNAP/") || path == "$SNAP" {
-		// NOTE: We use dirs.CoreSnapMountDir here as the path used will be always
-		// inside the mount namespace snap-confine creates and there we will
-		// always have a /snap directory available regardless if the system
-		// we're running on supports this or not.
-		// TODO parallel-install: use of proper instance/store name
-		return strings.Replace(path, "$SNAP", filepath.Join(dirs.CoreSnapMountDir, snapInfo.InstanceName(), snapInfo.Revision.String()), 1)
+	// Content cannot be mounted at arbitrary locations, expect the path to
+	// start with $SNAP, $SNAP_DATA, $SNAP_COMMON. If the prefix is none of
+	// these, place it under $SNAP.
```

Comment:
I appreciate the desire to keep this correct but I have a feeling we should have a validator that ensures only expected variables are used that is paired with very simple expansion logic, not the other way around. We have `snap.ValidatePathVariables` that checks for this already and we are using it for layouts. Should we just validate the essential snap path attributes more carefully?

Reply:
Maybe we could have ExpandSnapVariables() actually fail if unknown `$VAR` appears?

Reply:
Even if we use something that validates variables, the logic here is still special. The path here needs to _start_ at those locations.

That said, ValidatePathVariables looks nice indeed, and worth using here. I think we can just activate the first branch if ValidatePathVariables passes and the string starts with $. Otherwise, prepend $PATH and expand it as well. It's okay to kill non-existent variables for now. It's no different than the existing behavior, and not dangerous since we're expanding variables, cleaning the path, and prepending a safe path.

How does that sound?

Reply:
Ok, will do.

Reply:
Should be fixed now. Please take a look.